Fix mypy issues in fake_tensor.py

aorenste opened the pull request:
fake_tensor.py had mypy error ignored. That seems less than desirable.

Also added SafePyObjectT<T> which is a tagged wrapper around a SafePyObject but provides static type checking (with no other guarantees).

Used `SafePyObjectT<TorchDispatchModeKey>` on some of the TorchDispatchModeTLS API to ensure that we don't accidentally inject a different type than expected into the stack.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124771
* __->__ #124428

aorenste committed:
Fix mypy issues in fake_tensor.py

[ghstack-poisoned]

aorenste committed:
Update on "Fix mypy issues in fake_tensor.py"




[ghstack-poisoned]

aorenste committed:
Update on "Fix mypy issues in fake_tensor.py"




[ghstack-poisoned]

aorenste committed:
Update on "Fix mypy issues in fake_tensor.py"




[ghstack-poisoned]

aorenste committed:
Update on "Fix mypy issues in fake_tensor.py"




[ghstack-poisoned]

aorenste committed:
Update on "Fix mypy issues in fake_tensor.py"




[ghstack-poisoned]

aorenste committed:
Update on "Fix mypy issues in fake_tensor.py"

fake_tensor.py had mypy error ignored. That seems less than desirable.




[ghstack-poisoned]

malfet commented on .flake8:
```diff
@@ -13,6 +13,8 @@ ignore =
     EXE001,
     # these ignores are from flake8-bugbear; please fix!
     B007,B008,B017,B019,B023,B028,B903,B904,B905,B906,B907
+    # Coding magic comment not found - do we really want to add encoding to every file?
+    C101,
```
Please add codepage comments to those files instead (or strip the existing ones)
```suggestion
```

malfet reviewed (changes requested):
Please add codepage comment to the files you are adding (or strip all the existing ones)

malfet commented on .github/scripts/lintrunner.sh:
```diff
@@ -30,6 +30,9 @@ python3 -m tools.pyi.gen_pyi \
     --tags-path aten/src/ATen/native/tags.yaml \
     --deprecated-functions-path "tools/autograd/deprecated.yaml"
 
+# lintrunner will fail if the tee file exists
+rm -f lint.json
+
```
This sounds suspicious, how did it work all that time beforehand?

aorenste replied:
Most of the time it's fine because we run in a clean VM. But I had ssh'd to the CI machine debug to an issue and it had a stale file (and had really weird errors running this script since it failed but lintrunner said "no errors")

aorenste committed:
Update on "Fix mypy issues in fake_tensor.py"

fake_tensor.py had mypy error ignored. That seems less than desirable.




[ghstack-poisoned]

aorenste committed:
Update on "Fix mypy issues in fake_tensor.py"

fake_tensor.py had mypy error ignored. That seems less than desirable.




[ghstack-poisoned]

aorenste committed:
Update on "Fix mypy issues in fake_tensor.py"

fake_tensor.py had mypy error ignored. That seems less than desirable.




[ghstack-poisoned]

aorenste committed:
Update on "Fix mypy issues in fake_tensor.py"

fake_tensor.py had mypy error ignored. That seems less than desirable.




[ghstack-poisoned]

malfet commented on c10/core/SafePyObject.h:
```diff
@@ -55,6 +55,22 @@ struct C10_API SafePyObject {
+struct C10_API SafePyObjectT : private SafePyObject {
+  SafePyObjectT(PyObject* data, c10::impl::PyInterpreter* pyinterpreter)
+      : SafePyObject(data, pyinterpreter) {}
+  SafePyObjectT(SafePyObjectT&& other) noexcept : SafePyObject(other) {}
```
Curious, can't you just inherit them via `using SafePyObject::SafePyObject(SafePyObject&&)`?

aorenste replied:
I wanted to make sure that you couldn't accidentally turn an untyped SafePyObject into a typed SafePyObjectT.

malfet reviewed (approved):
LGTM, but don't you want to add newly covered files to `mypy.ini`/`mypy-strict.ini` or something? Otherwise what prevents people from introducing regressions again?

aorenste committed:
Update on "Fix mypy issues in fake_tensor.py"

fake_tensor.py had mypy error ignored. That seems less than desirable.

Also added SafePyObjectT<T> which is a tagged wrapper around a SafePyObject but provides static type checking (with no other guarantees).

Used `SafePyObjectT<TorchDispatchModeKey>` on some of the TorchDispatchModeTLS API to ensure that we don't accidentally inject a different type than expected into the stack.




[ghstack-poisoned]

aorenste committed:
Update on "Fix mypy issues in fake_tensor.py"

fake_tensor.py had mypy error ignored. That seems less than desirable.

Also added SafePyObjectT<T> which is a tagged wrapper around a SafePyObject but provides static type checking (with no other guarantees).

Used `SafePyObjectT<TorchDispatchModeKey>` on some of the TorchDispatchModeTLS API to ensure that we don't accidentally inject a different type than expected into the stack.




[ghstack-poisoned]

aorenste commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot revert -m "Unfortunately, I needed to revert #123735 and this one depends on it. So please check if there are no merge conflicts or breakages and feel free to merge this PR again" -c weird
